fixed arguments in mul_poly

srajangarg opened the pull request:
When we need to use the `RCP` of one of the arguments of the function, ( as required here) we cannot pass the reference to the class itself. This is because, we'll construct a stack variable, containing the required RCP of the argument. When the freeing of stack happens on the return of the function, it throws a `free() invalid pointer` error. As that memory location was not allocated on the heap.

This is also indicated by what certik said in https://github.com/symengine/symengine/pull/406#issuecomment-77592690

Sumith1896 commented:
+1 if tests pass.

isuruf commented on symengine/polynomial.cpp:
```diff
@@ -250,23 +250,23 @@ unsigned int bit_length(T t){
 
-RCP<const UnivariateIntPolynomial> mul_poly(RCP<const UnivariateIntPolynomial> a, RCP<const UnivariateIntPolynomial> b) {
-    //TODO: Use `const RCP<const UnivariateIntPolynomial> &a` for input arguments,
-    //      even better is use `const UnivariateIntPolynomial &a`
```
> even better is use `const UnivariateIntPolynomial &a`

srajangarg replied:
It's not possible to do so. Updating description.

Sumith1896 commented:
@isuruf Views?

isuruf commented:
I still don't get why `RCP` is needed. Where is it needed?

srajangarg commented:
You were right, I have managed to make the arguments `UnivariateIntPolynomial&`. Initially, the function was constructing a new negative polynomial on the heap (if it was required). This was redundant. I just had to add support for `max_coeff` handling the case if the whole polynomial was negative of what it currently is. This was done using a new parameter `negative` which defaults to `false`.

isuruf commented on symengine/polynomial.cpp:
```diff
@@ -130,11 +130,23 @@ vec_basic UnivariateIntPolynomial::get_args() const {
 
-integer_class UnivariateIntPolynomial::max_coef() const {
-    integer_class curr = dict_.begin()->second;
+integer_class UnivariateIntPolynomial::max_coef(bool negative) const {
```
Shouldn't this give the maximum absolute value coefficient?

srajangarg replied:
@Sumith1896 wrote it initially. But I believe it should be absolute value.

isuruf commented on symengine/polynomial.cpp:
```diff
@@ -274,8 +282,8 @@ RCP<const UnivariateIntPolynomial> mul_poly(RCP<const UnivariateIntPolynomial> a
     integer_class mask = a1 - 1;
-    integer_class sa = a->eval_bit(N);
-    integer_class sb = b->eval_bit(N);
+    integer_class sa = a_sign * a.eval_bit(N);
```
`a_sign` is not really needed, right?

srajangarg replied:
It is, according to how the old code was written.

isuruf replied:
Yes, but I don't think it is needed.

srajangarg replied:
Tests don't pass when I tried changing it. I think it requires the highest degree term to turn out positive maybe?

Sumith1896 commented:
Could you resolve the merge conflicts? The Polynomial PR was merged.

srajangarg commented:
Sure

certik commented:
Please do not mix styling changes with fixing bugs or implementing new features. It's really hard for me to review it, since most of the changes just move `{` around.

We should start using #866, then at least the formatting changes should go away.

srajangarg commented:
Sorry @certik, will keep in mind from now on.

srajangarg commented:
@certik @isuruf Does this require any work?

isuruf commented on symengine/polynomial.cpp:
```diff
@@ -174,19 +182,31 @@ integer_class UnivariateIntPolynomial::eval_bit(const int &x) const {
+{
+    if (dict_.empty())
+        return true;
+    return false;
```
Why this change?

isuruf commented:
@srajangarg, can you remove the styling changes?

srajangarg commented:
@isuruf done. Sorry about another commit, I had merged conflicts and made style changes together in the previous commit, so couldn't reset it.

Sumith1896 commented:
@isuruf How does it look to you? Shall I merge this?

isuruf commented:
@srajangarg, can you add test cases where there are negative coefficients and the largest absolute value coefficient is the constant term?

srajangarg commented:
@isuruf There was a huge mistake in `mul_poly`. The bound for N was incoorect!
The actual bound can be seen [here](http://www.cs.berkeley.edu/~fateman/papers/polysbyGMP.pdf#page=3) (page 3, para 1). I have corrected this.

I also added some test cases.

isuruf commented on symengine/tests/basic/test_polynomial.cpp:
```diff
@@ -76,14 +76,26 @@ TEST_CASE("Subtracting two UnivariateIntPolynomial", "[UnivariateIntPolynomial]"
-    RCP<const UnivariateIntPolynomial> a = univariate_int_polynomial(x, {{0, 1_z}, {1, 2_z}, {2, 1_z}});
-    RCP<const UnivariateIntPolynomial> b = univariate_int_polynomial(x, {{0, -1_z}, {1, -2_z}, {2, -1_z}});
+    map_uint_mpz adict_ = {{0, 1_z}, {1, 2_z}, {2, 1_z}};
+    map_uint_mpz bdict_ = {{0, -1_z}, {1, -2_z}, {2, -1_z}};
```
Can you add a test case like `{{0, -1},{1,-2},{2, -100}}`?

isuruf commented:
Thanks for looking into it. `mul_poly` needs lots more testing.

CodeMaxx commented:
@Sumith1896 I'm working on improving coverage for `polynomial.cpp` and some other files as well.

srajangarg commented:
@isuruf ping

isuruf commented on symengine/polynomial.cpp:
```diff
@@ -190,14 +190,16 @@ integer_class UnivariateIntPolynomial::eval_bit(const int &x) const
     for (auto it = dict_.rbegin(); it != dict_.rend(); ++it) {
 
-        result = (*it).second + (result << x * (last_deg - (*it).first));
+        two_pow = integer_class(1) << x * (last_deg - (*it).first);
```
Why this change?

isuruf commented on symengine/polynomial.cpp:
```diff
@@ -190,14 +190,16 @@ integer_class UnivariateIntPolynomial::eval_bit(const int &x) const
     }
-    result = result << x * last_deg;
+    two_pow = integer_class(1) << x * last_deg;
+    result *= two_pow;
```
Why this change?

isuruf commented on symengine/polynomial.cpp:
```diff
@@ -314,70 +316,48 @@ unsigned int bit_length(T t)
-        throw std::runtime_error("Error: variables must agree.");
-    } else {
-        var = a->get_var();
-    }
```
Why this change?

srajangarg replied:
This wasn't needed. Just need to check if the variables are equal or not, which I have added below.

isuruf replied:
It is needed. There can be polynomials with no var, just a coefficient which can be used later.
I think we need to rethink the design, but that's not related to this PR.

srajangarg replied:
So, what should I stick with now?

isuruf replied:
For now, keep the above as it is in the master, we'll figure it out later.

isuruf commented:
I think something went wrong while resolving conflicts. Can you check?

srajangarg commented:
@isuruf ping! Changes done from my side.

isuruf commented:
@srajangarg, thanks for the PR
